### PR TITLE
Do not reorder between copy_ and return statement

### LINKED ISSUE #912: `fuse_bound_symbols` (toposort on bsyms) can put `return` before `copy_` on arguments
## 🐛 Bug

[`fuse_bound_symbols`](https://github.com/Lightning-AI/lightning-thunder/blob/fa307a5656d2b7000a07a489f3d434aa7ba8cfd6/thunder/executors/data_dependent_partition.py#L300) (toposort on bsyms) puts `return` before `copy_` on arguments.

This causes AssertionError when `torchcompile` or `cudagraphex` executor is applied on in-place operations.

#### Code sample

```py
def f(a):
    a += 1

jitted = thunder.jit(f, executors=[thunder.get_executor("torchcompile")])
jitted(torch.tensor(0))
```

```
...
  File "/opt/pytorch/lightning-thunder/thunder/common.py", line 873, in transform_to_torch_types
    assert last.sym.id == prims.PrimIDs.RETURN
AssertionError
```

Before reaching this line, `TorchCompileExecution` generates
```py
# Constructed by Transform for execution (took 3 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a):
  # a: "cpu i64[]"

  # /opt/pytorch/lightning-thunder/mshinokawa/sandbox/debug.py:31: 	    a += 1
  return None
  TorchCompile0(a)
    # t0 = ltorch.add(a, 1, alpha=None)  # t0: "cpu i64[]"
      # t0 = prims.add(a, 1)  # t0: "cpu i64[]"
    # prims.copy_(t0, a)
```

Returning `a` at the end of `f` does not fix this.

```py
# Constructed by Transform for execution (took 3 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a):
  # a: "cpu i64[]"

  # /opt/pytorch/lightning-thunder/mshinokawa/sandbox/debug.py:5: 	    a += 1
  return a
  TorchCompile0(a)
    # t0 = ltorch.add(a, 1, alpha=None)  # t0: "cpu i64[]"
      # t0 = prims.add(a, 1)  # t0: "cpu i64[]"
    # prims.copy_(t0, a)
```

## Cause

Inside `fuse_bound_symbols`, [`Graph.__init__`](https://github.com/Lightning-AI/lightning-thunder/blob/fa307a5656d2b7000a07a489f3d434aa7ba8cfd6/thunder/executors/data_dependent_partition.py#L80) constructs a dependency graph and apply topological sort to reorder bound symbols. The dependency is found via `bsym.flat_args` and `bsym.flat_outs`, and `prims.copy_` has no output. Hence the algorithm does not spot the dependency between `copy_(t0, a)` and `return a`.

Although `nvfuser` uses this function too, it gets around this by forcefully moving the return statement after all the other bound symbols ([here](https://github.com/Lightning-AI/lightning-thunder/blob/fa307a5656d2b7000a07a489f3d434aa7ba8cfd6/thunder/executors/nvfuserex_impl.py#L821-L832)).

## Related issue

#229 suggested adding `TorchCompileExecutor` to the tests along with `TorchExecutor` and `nvFuserExecutor`.

### PR DESCRIPTION
**Pull request by shino16**
## What does this PR do?

Fixes #912. Registers `COPY_` symbols before the return statement as its dependencies.

### CODE REVIEW

File: thunder/executors/data_dependent_partition.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ def __init__(self, trace: TraceCtx):
         # as it appears to be far off from being universal.
         # We use indices as hash values instead.
         bsym_id_to_node_map: list[int] = []
+        copy_nodes: list[Node] = []
```
Comment (t-vi) [inline thunder/executors/data_dependent_partition.py:92]: Would it be advantageous to use "fast to lookup" data structure here if we have a membership test below?

**Review by t-vi — COMMENTED** [review]
Hi @shino16 ! 
Looks great, congrats!
I have one comment regarding the copy_nodes, but other than that I think it's all good for merging.

**Comment by shino16** [comment]
`torch.compile` is currently not yet available on Windows: https://github.com/pytorch/pytorch/issues/124245#issuecomment-2267103009

File: thunder/executors/data_dependent_partition.py (hunk 1 above)
Comment (t-vi) [reply thunder/executors/data_dependent_partition.py:92]: Oh, right. I saw the in and thought if... Looking at PRs before coffee.

**Review by t-vi — APPROVED** [review]
Great analysis and patch, thank you @shino16

**Pull request merged**